lib/charms/opensearch/v0/opensearch_health.py

```diff
@@ -58,42 +61,87 @@ def apply(
+
+        # the health depends on data nodes, for large deployments: an ML cluster
+        # may not be concerned about reporting or relying on the health of the
+        # data nodes in other clusters. We should therefore get this info from
+        # the deployment descriptor which has an overview of all the cluster.
+        # compute health only in clusters where data nodes exist
+        compute_health = (
+            deployment_desc.start == StartMode.WITH_GENERATED_ROLES
+            or "data" in deployment_desc.config.roles
+            or not local_app_only
+        )
+        if not compute_health:
+            return HealthColors.IGNORE
+
+        host = self._charm.unit_ip if use_localhost else None
+        response = self._health(host, wait_for_green_first)
+        if wait_for_green_first and not response:
+            response = self._health(host, False)
+
+        if not response:
+            return HealthColors.UNKNOWN
 
-            if app:
-                self.apply_for_app(status)
-            else:
-                self.apply_for_unit(status)
+        logger.info(f"Health: {response}")
+        try:
+            status = response["status"].lower()
+        except AttributeError as e:
+            logger.error(e)  # means the status was reported as an int (i.e: 503)
+            return HealthColors.UNKNOWN
 
+        if status != HealthColors.YELLOW:
             return status
+
+        try:
+            logger.debug(
+                f"\n\nHealth: {status} -- Shards: {ClusterState.shards(self._opensearch, host)}\n\n"
+            )
         except OpenSearchHttpError:
-            return HealthColors.UNKNOWN
+            pass
+
+        # we differentiate between a temp yellow (moving shards) and a permanent
+        # one (such as: missing replicas)
+        if response["initializing_shards"] > 0 or response["relocating_shards"] > 0:
+            return HealthColors.YELLOW_TEMP
+        return HealthColors.YELLOW
+
+    @retry(stop=stop_after_attempt(15), wait=wait_fixed(5), reraise=True)
+    def wait_for_shards_relocation(self) -> None:
+        """Blocking function until the shards relocation completes in the cluster."""
+        if self.get(wait_for_green_first=True) != HealthColors.YELLOW_TEMP:
+            return
 
-    def apply_for_app(self, status: str) -> None:
+        # we throw an error because various operations should NOT start while data
+        # is being relocated. Examples are: simple stop, unit removal, upgrade
+        raise OpenSearchHAError("Shards haven't completed relocating.")
+
+    def _apply_for_app(self, status: str) -> None:
         """Cluster wide / app status."""
         if not self._charm.unit.is_leader():
-            # this is needed in case the leader is in an error state and doesn't
-            # report the status itself
-            self._charm.peers_data.put(Scope.UNIT, "health", status)
+            trigger_peer_rel_changed(self._charm, on_other_units=True)
```

Comment:
If we will apply "something" in the peer relation, why not applying the `status`?

Reply:
Actually, this is never going to be called:
`_apply_for_app` only runs for app-level, this is supposed to never run at app-level (not leader).

Reply:
Because the status is a variable that changes with time, setting/getting it from the peer rel data may lead to outdated status reporting (we used to do it like this on scale-down, it didn't work well)

Reply:
@Mehdi-Bendriss not sure if you saw my 2nd comment here:

> Actually, this is never going to be called:
_apply_for_app only runs for app-level, this is supposed to never run at app-level (not leader).

(It was 1 min apart from yours)

In a nutshell, if I understood correctly that should never be called.